core/src/main/java/hudson/security/HudsonPrivateSecurityRealm.java

```diff
@@ -717,14 +717,21 @@ public Details newInstance(StaplerRequest req, JSONObject formData) throws FormE
                 if(!Util.fixNull(pwd).equals(Util.fixNull(pwd2)))
                     throw new FormException("Please confirm the password by typing it twice","user.password2");
 
+                User user = Util.getNearestAncestorOfTypeOrThrow(req, User.class);
+                UserProperty p = user.getProperty(Details.class);
+                String currentUserHashedPassword = ((Details) p).getPassword();
+
                 String data = Protector.unprotect(pwd);
                 if(data!=null) {
                     String prefix = Stapler.getCurrentRequest().getSession().getId() + ':';
                     if(data.startsWith(prefix))
+                        PasswordPropertyListener.fireOnChanged(user.getId(), Util.fixNull(currentUserHashedPassword), Util.fixNull(pwd));
                         return Details.fromHashedPassword(data.substring(prefix.length()));
                 }
 
-                User user = Util.getNearestAncestorOfTypeOrThrow(req, User.class);
+                if (p != null) {
```

Comment:
You are checking here if p is not null, but you already called the getPassword() method a few lines before.